implementation of IMap.valuesAsync() and IMap.valuesAsync(Predicate)

jeeminso opened the pull request:
attempt to resolve  #17370 .

currently, QueryEngine.execute invokes executeAsync which I have implemented.

Tests passed (MapAggregateTest, MapValuesTest, QueryEngineImplTest)

Please provide a feedback whether I am on the right track.

Also, the original design is to used partitionIdSet as a shared state to keep the ids that need re-dispatching. (within addResultsOfPredicate)

I thought I could remove that and make it more async, but was not sure if there will be any benefit from it, so I kept it.

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org/pull/badge/not_signed)](https://cla.hazelcast.org/hazelcast/hazelcast?pullRequest=18101) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.hazelcast.org/hazelcast/hazelcast?pullRequest=18101) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org/check/hazelcast/hazelcast?pullRequest=18101) it.</sub>

jeeminso commented:
@mmedenjak hello, could I ask what is the typical process for a review? I was not sure If I could just wait or I am missing something

mmedenjak commented:
run-lab-run

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/client/impl/ClientDelegatingFuture_SerializationExceptionTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/spi/impl/DelegatingCompletableFuture_SerializationExceptionTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
</pre>
</details>

mmedenjak commented:
@jeeminso thank you for your contribution! I'll try to get someone to review it. Now, you have checkstyle issues so can you run `mvn clean package -Pcheckstyle,spotbugs -DskipTests` locally and fix those? Also, make sure you run it every time before you push changes, to avoid situations like this one.

taburet commented on hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryEngineImpl.java:
```diff
@@ -172,47 +187,61 @@ private Result populateResult(Query query, PartitionIdSet partitionIds) {
+                                // see also https://github.com/hazelcast/hazelcast/issues/6471
+                                continue;
+                            }
+                            finishedPartitionIds.removeAll(queriedPartitionIds);
```
It might be a good time to give a new name to `finishedPartitionIds`, to my understanding the usage is opposite to the naming, it's like "unfinished" or "left" partitions.

taburet commented:
The direction looks good. The only concern is the increased number of allocations leading to the increased GC pressure. While in some places it's hard to avoid it, that's how Java APIs work, in some other places it might easily avoided. For instance, in `addResultsOfPredicate` the array passed to `CompletableFuture.allOf` might be constructed directly in a classic old-school way to avoid intermediary `List` instance (which is usually the regular `ArrayList` under the hood).

jeeminso commented:
thanks for your feedback, I will fix the two issues and handle the remaining portion.

jeeminso commented:
I have implemented IMAP methods and added a few tests. (a few since non-async methods now simply invokes async versions with ```get()``` or its variants)

I am having trouble with the spotbug script ```mvn clean package -Pcheckstyle,spotbugs -DskipTests``` Please help get around these errors:
```
[ERROR] Medium: Load of known null value in com.hazelcast.internal.util.phonehome.BuildInfoCollector.getDownloadId() [com.hazelcast.internal.util.phonehome.BuildInfoCollector] At BuildInfoCollector.java:[line 58] NP_LOAD_OF_KNOWN_NULL_VALUE
[ERROR] Medium: Redundant nullcheck of is, which is known to be non-null in com.hazelcast.internal.util.phonehome.BuildInfoCollector.getDownloadId() [com.hazelcast.internal.util.phonehome.BuildInfoCollector] Redundant null check at BuildInfoCollector.java:[line 51] RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE
[ERROR] Medium: Redundant nullcheck of is which is known to be null in com.hazelcast.internal.util.phonehome.BuildInfoCollector.getDownloadId() [com.hazelcast.internal.util.phonehome.BuildInfoCollector] Redundant null check at BuildInfoCollector.java:[line 58] RCN_REDUNDANT_NULLCHECK_OF_NULL_VALUE
```

jeeminso commented:
could someone have a look? I have manually merged. Thanks
